server/src/main/java/org/apache/druid/metadata/segment/cache/HeapMemorySegmentMetadataCache.java

```diff
@@ -0,0 +1,623 @@
+  }
+
+  /**
+   * Verifies that the cache is ready to serve requests, waiting if necessary.
+   *
+   * @throws DruidException if the cache is disabled, stopped or not leader.
+   */
+  private void verifyCacheIsReady()
+  {
+    if (!isCacheEnabled) {
+      throw DruidException.defensive("Segment metadata cache is not enabled.");
+    }
+
+    synchronized (cacheStateLock) {
+      switch (currentCacheState) {
+        case STOPPED:
+          throw DruidException.defensive("Segment metadata cache has not been started yet.");
+        case FOLLOWER:
+          throw DruidException.defensive("Not leader yet. Segment metadata cache is not usable.");
+        case LEADER_FIRST_SYNC_PENDING:
+        case LEADER_FIRST_SYNC_STARTED:
+          waitForCacheToFinishSync();
+          verifyCacheIsReady();
+        case LEADER_READY:
+          // Cache is now ready for use
+      }
+    }
+  }
+
+  /**
+   * Waits for cache to become ready if we are leader and current state is
+   * {@link CacheState#LEADER_FIRST_SYNC_PENDING} or
+   * {@link CacheState#LEADER_FIRST_SYNC_STARTED}.
+   */
+  private void waitForCacheToFinishSync()
+  {
+    synchronized (cacheStateLock) {
+      log.info("Waiting for cache to finish sync with metadata store.");
+      while (currentCacheState == CacheState.LEADER_FIRST_SYNC_PENDING
+             || currentCacheState == CacheState.LEADER_FIRST_SYNC_STARTED) {
+        try {
+          cacheStateLock.wait(5 * 60_000);
+        }
+        catch (Exception e) {
+          log.noStackTrace().error(e, "Error while waiting for cache to be ready");
+          throw DruidException.defensive(e, "Error while waiting for cache to be ready");
+        }
+      }
+      log.info("Wait complete. Cache is now in state[%s].", currentCacheState);
+    }
+  }
+
+  private void markCacheAsReadyIfLeader()
+  {
+    synchronized (cacheStateLock) {
+      if (currentCacheState == CacheState.LEADER_FIRST_SYNC_STARTED) {
+        currentCacheState = CacheState.LEADER_READY;
+        log.info("Sync has finished. Cache is now ready to serve requests.");
+
+        // Notify waiting threads waiting for cache to be ready
+        cacheStateLock.notifyAll();
+      }
+    }
+  }
+
+  /**
+   * Retrieves segments from the metadata store and updates the cache, if required.
+   * <p>
+   * The following actions are performed in every sync:
+   * <ul>
+   * <li>Retrieve all used and unused segment IDs along with their updated timestamps</li>
+   * <li>Retrieve payloads of used segments which have been updated in the metadata
+   * store but not in the cache</li>
+   * <li>Retrieve all pending segments and update the cache as needed</li>
+   * <li>Remove segments not present in the metadata store</li>
+   * <li>Reset the max unused partition IDs</li>
+   * <li>Change the cache state to ready if it is leader and waiting for first sync</li>
+   * <li>Emit metrics</li>
+   * </ul>
+   */
+  private void syncWithMetadataStore()
```

Comment:
Have you been able to benchmark this method with a cluster with lots of segments (e.g. millions)? It will need to complete before allocation can work post-leadership-election, so I'm wondering how much time that will take.

Reply:
Not yet. I am about to start testing on some clusters with a large number of segments. I will share the results here once the testing is done.

Reply:
Ran some tests on a local cluster with 600K segments.
Full sync takes about 1 min.
Delta sync (fetch all segment IDs but payloads only for updated segments) takes about 2s.

So, after an Overlord is elected leader, it should be able to sync up within a few seconds.
I will do some more testing involving leader re-elections.